### ISSUE TITLE
Is it still possible to create nodes based on a GraphQL query?

### ISSUE THREAD

**Issue by fabe** [gatsbyjs/gatsby#16150]
Up until recently I was able to create nodes based on a GraphQL query. For example in my starter [gatsby-starter-deck](https://github.com/fabe/gatsby-starter-deck), I created nodes based on generated HTML via `gatsby-transformer-remark`. For every section, a page and a node was created (to benefit from hot reloading):

```js
exports.createPages = ({ actions, createNodeId, graphql }) => {
  const { createPage, createNode } = actions;

  return graphql(`
    {
      allMarkdownRemark {
        ...
      }
    }
  `).then(({ nodes }) => {
    nodes.forEach(({ node, html }, index) => {
      const splitNodes = html.split('<hr>');

      splitNodes.forEach(n => createNode({ ... }));

      // Then create page that queries the generated node.
    });
  });
};
```

This does not work anymore since ~`gatsby@2.2` (fabe/gatsby-starter-deck#17).

**Is this still possible at all or do I need to either leave out hot reloading (by passing the split HTML via page context) or handle the splitting on the client (passing all HTML to every page)?**

### Environment (if relevant)

```bash
  System:
    OS: macOS 10.14.5
    CPU: (8) x64 Intel(R) Core(TM) i7-6700HQ CPU @ 2.60GHz
    Shell: 5.3 - /bin/zsh
  Binaries:
    Node: 11.9.0 - /usr/local/bin/node
    Yarn: 1.13.0 - /usr/local/bin/yarn
    npm: 6.5.0 - /usr/local/bin/npm
  Languages:
    Python: 2.7.10 - /usr/bin/python
  Browsers:
    Chrome: 75.0.3770.142
    Safari: 12.1.1
  npmPackages:
    gatsby: ^2.0.33 => 2.0.33
    gatsby-plugin-layout: ^1.0.1 => 1.0.1
    gatsby-plugin-offline: ^2.0.5 => 2.0.5
    gatsby-plugin-react-helmet: ^3.0.0 => 3.0.0
    gatsby-plugin-sharp: ^2.0.14 => 2.0.14
    gatsby-remark-copy-images: ^0.2.1 => 0.2.1
    gatsby-remark-copy-linked-files: ^2.0.7 => 2.0.7
    gatsby-remark-images: ^3.0.1 => 3.0.1
    gatsby-remark-relative-images: ^0.2.1 => 0.2.1
    gatsby-source-filesystem: ^2.0.1 => 2.0.1
    gatsby-transformer-remark: ^2.1.15 => 2.1.15
    gatsby-transformer-sharp: ^2.1.9 => 2.1.9
  npmGlobalPackages:
    gatsby-cli: 2.7.20
```

It breaks with `gatsby@2.2.*`.

### File contents (if changed)

See https://github.com/fabe/gatsby-starter-deck/blob/master/gatsby-node.js

**Comment by stefanprobst** [comment]
Before Gatsby 2.2, we created a GraphQL schema twice -- in the "build schema" step, and again in the "update schema" step. The latter was mainly introduced to have fields added to `SitePage.context` in the `createPages` API available in the schema.

With the introduction of the schema customization APIs in 2.2, we limited the "update schema" step to run type inference only on `SitePage` nodes, to avoid having to do everything twice. Unfortunately, this broke scenarios like in your starter.

However, this can be fixed by explicitly defining the `Slide` type upfront with the `createTypes` action (added in 2.2). See https://github.com/fabe/gatsby-starter-deck/blob/50a3ab372f76bc735d1757853fe81207739d1afd/gatsby-node.js#L79-L86
